pkg/controller/clusterrelocate/clusterrelocate_controller.go

```diff
@@ -0,0 +1,812 @@
+package clusterrelocate
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	log "github.com/sirupsen/logrus"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/diff"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
+	hivemetrics "github.com/openshift/hive/pkg/controller/metrics"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+	"github.com/openshift/hive/pkg/remoteclient"
+)
+
+const (
+	ControllerName = "clusterRelocate"
+)
+
+var (
+	typesToCopy = func() []runtime.Object {
+		return []runtime.Object{
+			&corev1.SecretList{},
+			&corev1.ConfigMapList{},
+			&hivev1.MachinePoolList{},
+			&hivev1.SyncSetList{},
+			&hivev1.SyncIdentityProviderList{},
+		}
+	}
+
+	metricSuccessfulClusterRelocations = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "hive_cluster_relocations",
+		Help: "Total number of successful cluster relocations.",
+	}, []string{"cluster_relocate"})
+
+	metricAbortedClusterRelocations = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "hive_aborted_cluster_relocations",
+		Help: "Total number of aborted cluster relocations.",
+	}, []string{"cluster_relocate", "reason"})
```

Comment:
I believe both of these metrics should be counters, not gauges. Gauge would be for a value reflecting a current state, in this case it looks like you'd get the number of relocations since the pod was restarted. Counter will give us an ever incrementing value and allow the use of rate.

Reply:
Definitely!